### ISSUE TITLE
2.0 Roadmap

### ISSUE THREAD

**Issue by lilleyse** [AnalyticalGraphicsInc/gltf-pipeline#330]
**Background:**

The [2.0 branch](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/tree/2.0) is used by Cesium to upgrade 1.0 models to 2.0. The upgraded models use the [KHR_technique_webgl](https://github.com/KhronosGroup/glTF/tree/master/extensions/Khronos/KHR_technique_webgl) extension, which is supported by Cesium but hardly any other engines. PBR is not supported in this branch.

There is another branch called [cleanup](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/tree/cleanup) which is a cleanup of glTF-pipeline that supports 1.0-2.0 conversion, conversions between glTF and glb, and other basic tasks. The other pipeline stages will eventually be moved over here once the core is solid.

The `2.0` branch will be renamed to `2.0-cesium` and the `cleanup` branch renamed to `2.0`.

One confusing aspect of the project right now is a subset of the files are used within Cesium for upgrading 1.0 models. Cesium includes gltf-pipeline which includes Cesium. Our early attempts at using browserify didn't work, so we instead added a gulp task called `build-cesium` which saves those gltf-pipeline files with AMD syntax to be packaged directly with Cesium. Additionally those files do not use the Node.js API or any third party packages (except Cesium). If anyone can think of an alternative to this approach that would be very helpful.

Another side effect of including gltf-pipeline in Cesium is we try to make those files as performant as possible. We try to avoid any large buffer copies, and because of that glTF 1.0 models that are upgraded to 2.0 are not always byte aligned correctly. This is best for Cesium but not best for gltf-pipeline, so there needs to be some compromise here.

**To do:**

* [ ] Bring the new `2.0` branch up to good enough quality so others can start working on it. It is almost there.
* [ ] `updateVersion` needs to fix accessors that have invalid byte alignment: https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#data-alignment
* [ ] Support upgrading glTF 1.0 materials to PBR in `updateVersion`. This means removing shaders and techniques and making some best guesses for upgrading to PBR.
* [ ] A lot of models are using the old `KHR_materials_common` extension. There should be a stage to convert this to PBR.
* [ ] Transfer all stages over.
* [ ] Support exporting with specular-glossiness. This will require a metallic-roughness to specular-glossiness converter.
* [ ] Continue to support `KHR_technique_webgl` throughout the pipeline since having custom shaders is very important for some of Cesium's users.

**Comment by pjcozzi** [comment]
>  Bring the new 2.0 branch up to good enough quality so others can start working on it. It is almost there.

I'm going to wait on this before we start to ask for more contributors.

**Comment by pjcozzi** [comment]
> Another side effect of including gltf-pipeline in Cesium is we try to make those files as performant as possible. We try to avoid any large buffer copies, and because of that glTF 1.0 models that are upgraded to 2.0 are not always byte aligned correctly. This is best for Cesium but not best for gltf-pipeline, so there needs to be some compromise here.

gltf-pipeline should be conformant to the glTF spec.  We should make the fast runtime engine conversion an option that defaults to off.

**Comment by lilleyse** [comment]
> gltf-pipeline should be conformant to the glTF spec. We should make the fast runtime engine conversion an option that defaults to off.

Yeah I agree on that, we already use an `optimizeForCesium` flag elsewhere that we can use here too.

**Comment by donmccurdy** [comment]
> * [ ] Support exporting with specular-glossiness. This will require a metallic-roughness to specular-glossiness converter.

I think this could be filed separately as a post-2.0 enhancement. Great if someone wants to take it, but not launch-blocking.

> * [ ] Continue to support KHR_technique_webgl throughout the pipeline since having custom shaders is very important for some of Cesium's users.

Agreed, but let's make this an `--experimental_khr_technique_webgl` flag, that is **off** by default, to avoid having these models spread too far before the extension is finalized.

> We try to avoid any large buffer copies, and because of that glTF 1.0 models that are upgraded to 2.0 are not always byte aligned correctly. 

Not sure I understand — is this something that should have an enable/disable flag, or better to just debug it if/when we find issues?

**Comment by donmccurdy** [comment]
> One confusing aspect of the project right now is a subset of the files are used within Cesium for upgrading 1.0 models. Cesium includes gltf-pipeline which includes Cesium. Our early attempts at using browserify didn't work...

Not quite following... do you mean there are files missing from gltf-pipeline that are provided by Cesium, so running `browserify index.js -o bundle.js` doesn't work? What I see when trying that now is something like:

```
Error: Cannot find module './lib/AccessorReader' from '/Users/donmccurdy/Documents/Projects/gltf-pipeline'
```

**Comment by lilleyse** [comment]
> I think this could be filed separately as a post-2.0 enhancement. Great if someone wants to take it, but not blocking 2.0 release.

Sounds good, I moved it to a "Later" section.

> Agreed, but let's make this an --experimental_khr_technique_webgl flag, that is off by default, to avoid having these models spread too far before the extension is finalized.

Also sounds good.

> Not sure I understand — is this something that should have an enable/disable flag, or better to just debug it if/when we find issues?

A lot of glTF 1 models do not have proper byte alignment for their accessors, so in order to create a TypedArray of an unaligned accessor the data must be copied to a new ArrayBuffer first. We try to avoid doing this in Cesium to avoid that overhead. And since some gltf-pipeline files are included in Cesium we try to avoid any copies there too.

But after looking at the glTF 1 spec I realized byte alignment was required the whole time. I thought it was only added for glTF 2. This means the models have never been "correct" and maybe we shouldn't be worrying as much about the copy overhead. Instead Cesium should just report a warning if it notices unaligned accessors. So what I wrote above probably doesn't matter as much, we should handle this situation however is best for gltf-pipeline, not Cesium. I don't think any flags are needed.

> Not quite following... do you mean there are files missing from gltf-pipeline that are provided by Cesium, so running browserify index.js -o bundle.js doesn't work? What I see when trying that now is something like:

Basically we wanted to be able to use gltf-pipeline's 1.0-2.0 converter in Cesium. Usually browserify or webpack are they way to go here, (and I forget the details) but we weren't able to get it to work. I think part of the problem was that gltf-pipeline includes the Cesium npm package. We included Cesium in order to get some of its utility functions and normals-generation code. But if Cesium the JS library includes a browserified gltf-pipeline, which includes the Cesium npm package, some circular dependency problem starts to happen.

So I guess the basic question is, how can we bring gltf-pipeline to the web to be used by Cesium and other engines? Our somewhat hacky solution was to amd-ify a subset of the files. The major downside here was we could no longer use the Node.js API in those files. So in the code you might notice that we're using ArrayBuffers/TypedArrays in some places and Buffers in others. This was meant to be a short term solution from a start, so I would really like to see this approach changed before the next version of gltf-pipeline is considered ready. By the way, I haven't tried browserify myself so I might be able to experiment a little.

**Comment by lasalvavida** [comment]
So, this problem has a few parts:

1. **Cesium uses requirejs in order to provide an npm package**
    Because Cesium uses AMD for its modules, the endpoint that is exposed is actually an instance of requirejs pointed at the Cesium source tree. This barrier made it impossible by any means that I explored to resolve Cesium dependencies in gltf-pipeline with a bundler.
    If you look at my [node-cesium-experimental](https://github.com/lasalvavida/node-cesium-experimental) repository, that was my proposed solution to this problem, which to my knowledge does work. It reformats and conditions the Cesium source code from AMD -> CommonJS and exposes Cesium sources directly from index.js so that Browserify and Webpack can read it.
    However, this would have been a big change to Cesium so we decided that in order to get glTF 2.0 support in Cesium moving faster that we would have to work around this. So instead of bringing Cesium to gltf-pipeline, we brought gltf-pipeline to Cesium.

2. **ES5 CommonJS require statements can't be optimized by any existing bundler**
    Simple put: `var Cesium = require('cesium'); var Cartesian3 = Cesium.Cartesian3;` can't be optimized to only bundle `Cartesian3`. This is usually called "tree-shaking" by bundlers. Webpack and Roll-up do support this for ES6 imports: `import Cartesian3 from 'cesium'`. But since we weren't even willing to move Cesium from AMD -> ES5, I didn't explore going from ES5 -> ES6 too deeply. I think there are some babel based projects that attempt this though. The alternative that I settled on for the work I did do in my experimental repo was splitting Cesium into smaller packages, `Core`, `Renderer`, etc, and then having gltf-pipeline depend on a subset of Cesium.

tl;dr: In order for gltf-pipeline to be bundled cleanly, it either has to stop depending on Cesium, or Cesium has to change how it exports its modules.

**Comment by lilleyse** [comment]
Thanks for the clarifications @lasalvavida - that is a great summary. I know ES6 modules is on the Cesium roadmap, but I wouldn't want that to hold up gltf-pipeline. I think the best solution moving forward is to keep the `2.0-cesium` branch in a working state for inclusion in Cesium, but treat the `2.0` branch as its own thing and not worry about how it interacts with Cesium as a third-party. This will make things a lot simpler in the long run.

**Comment by webprofusion-chrisc** [comment]
Nobody wants a "me too!" comment, but this is a pretty big deal now that facebook embed .glb. My files are gltf and I don't seem to have a way to convert them from there to .glb using node.

**Comment by lilleyse** [comment]
@webprofusion-chrisc you can try out the [2.0 branch](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/tree/2.0).

It's not that thoroughly tested but it supports:
* Converting glTF to glb (and reverse)
* Saving buffers/textures as embedded or separate files
* Converting glTF 1.0 models to glTF 2.0 (using the [KHR_technique_webgl extension](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_technique_webgl))

**Comment by donmccurdy** [comment]
@webprofusion-chrisc FWIW i'm using https://sbtron.github.io/makeglb/ for glTF->GLB with some success. The Facebook importer is extremely strict about validation (probably for the best), so if there are any validation errors, so may want to run the model through validation [here](https://gltf-viewer.donmccurdy.com/) or [here](http://github.khronos.org/glTF-Validator/) first.

**Comment by webprofusion-chrisc** [comment]
@lilleyse this is great, it does indeed convert my gltf! The resulting .glb gives my a few validation messages:
```json
"messages": [
            {
                "code": "NODE_MATRIX_DEFAULT",
                "message": "Do not specify default transform matrix.",
                "severity": 2,
                "pointer": "/nodes/0/matrix"
            },
            {
                "code": "NODE_MATRIX_DEFAULT",
                "message": "Do not specify default transform matrix.",
                "severity": 2,
                "pointer": "/nodes/2/matrix"
            },
....
...
````

@donmccurdy thanks, yes I found makeglb after some googling and it seemed OK but has no options to tweak. Your viewer is great and my new glb (exported from a threejs scene and run through gltf-pipeline 2.0) is working fine with no issues. I had to tweak processBufferView in the threejs exporter to compensate for 4-byte alignment again, I'll do a PR. :
![glb](https://user-images.githubusercontent.com/2445502/36625461-df26a056-195a-11e8-89fc-5ac339137e9e.png)

Now if I could just get facebook to allow up to 6MB instead of 3MB. Our models were originally >40MB so to get it down to 5, with textures, was a stretch.

**Comment by donmccurdy** [comment]
dev version of GLTFExporter should have the other 4-byte alignment issue ironed out now, I think.

**Comment by lilleyse** [comment]
Thanks for the notes @webprofusion-chrisc, the `NODE_MATRIX_DEFAULT` problem is fixed.

**Comment by lilleyse** [comment]
@TimvanScherpenzeel We plan on removing compressed texture support in the `2.0` branch. It wasn't widely used and like you pointed out it would be better to wait for a glTF 2 extension.

**Comment by ggetz** [comment]
Moved "Future" items to their own issues with [post 2.0](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/labels/post%202.0) label

**Comment by lilleyse** [comment]
@donmccurdy the 2.0 branch is really close to getting merged into master. Hopefully sometime this week. After we'll publish to npm.

**Comment by donmccurdy** [comment]
Sounds great, thanks!

**Comment by lilleyse** [comment]
2.0 is now merged into master and publish to npm!
